lib/rex/io/stream_server.rb

```diff
@@ -1,228 +1,208 @@
+          monitor_clients
+        end
+      end
 
-        # Append to the list of clients
-        self.clients << cli
+      #
+      # Terminates the listener monitoring threads and closes all active clients.
+      #
+      def stop
+        listener_thread.kill
+        clients_thread.kill
 
-        # Initialize the connection processing
-        on_client_connect(cli)
+        clients.each do |cli|
+          close_client(cli)
+        end
+      end
 
-        # Notify the client monitor
-        self.client_waiter.push(cli)
+      #
+      # This method closes a client connection and cleans up the resources
+      # associated with it.
+      #
+      def close_client(client)
+        if client
+          detach_client(client)
+
+          begin
+            client.close
+          rescue IOError
+          end
+        end
+      end
 
-      # Skip exceptions caused by accept() [ SSL ]
-      rescue ::EOFError, ::Errno::ECONNRESET, ::Errno::ENOTCONN, ::Errno::ECONNABORTED
-      rescue ::Interrupt
-        raise $!
-      rescue ::Exception
-        elog("Error in stream server server monitor: #{$!}")
-        rlog(ExceptionCallStack)
-        break
+      #
+      # Detach a client. You are now responsible for it, not us
+      #
+      def detach_client(client)
+        clients.delete(client)
       end
-    end
-  end
 
-  #
-  # This method monitors client connections for data and calls the
-  # +on_client_data+ routine when new data arrives.
-  #
-  def monitor_clients
-    begin
-
-      # Wait for a notify if our client list is empty
-      if (clients.length == 0)
-        self.client_waiter.pop
-        next
+      #
+      # This method waits on the server listener thread
+      #
+      def wait
+        listener_thread.join if listener_thread
       end
 
-      sd = Rex::ThreadSafe.select(clients, nil, nil, nil)
+      ##
+      #
+      # Callback procedures.
+      #
+      ##
+
+      #
+      # This callback procedure can be set and will be called when new clients
+      # connect.
+      #
+      attr_accessor :on_client_connect_proc
+      #
+      # This callback procedure can be set and will be called when clients
+      # have data to be processed.
+      #
+      attr_accessor :on_client_data_proc
+      #
+      # This callback procedure can be set and will be called when a client
+      # disconnects from the server.
+      #
+      attr_accessor :on_client_close_proc
+
+      attr_accessor :clients, :listener_thread, :clients_thread, :client_waiter # :nodoc: # :nodoc:
+
+      protected
+
+      #
+      # This method monitors the listener socket for new connections and calls
+      # the +on_client_connect+ callback routine.
+      #
+      def monitor_listener
+        while true
+          begin
+            cli = accept
+            unless cli
+              elog('The accept() returned nil in stream server listener monitor')
```

Comment:
Dropped the `fd.inspect` part because `fd` isn't available in Meterpreter channels which caused an exception to be thrown.